File: src/main/java/com/mingliqiye/utils/json/GsonJsonApi.java
package com.mingliqiye.utils.json;

import com.google.gson.*;

public class GsonJsonApi implements JsonApi {

	private final Gson gson;
	private final Gson gsonUnicode;
	private final Gson gsonPretty;
	private final Gson gsonPrettyUnicode;

	public GsonJsonApi() {
		gson = new GsonBuilder()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();

		gsonUnicode = new GsonBuilder()
			.disableHtmlEscaping()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();

		gsonPretty = new GsonBuilder()
			.setPrettyPrinting()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();

		gsonPrettyUnicode = new GsonBuilder()
			.setPrettyPrinting()
			.disableHtmlEscaping()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();
	}

	public GsonJsonApi(Gson gson) {
		this.gson = gson
			.newBuilder()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();
		this.gsonUnicode = gson
			.newBuilder()
			.disableHtmlEscaping()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();
		this.gsonPretty = gson
			.newBuilder()
			.setPrettyPrinting()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();
		this.gsonPrettyUnicode = gson
			.newBuilder()
			.setPrettyPrinting()
			.disableHtmlEscaping()
			.setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
			.create();
	}

	@Override
	public <T> T parse(String json, Class<T> clazz) {
		return gson.fromJson(json, clazz);
	}

	@Override
	public <T> T parse(String json, JsonTypeReference<T> type) {
		return gson.fromJson(json, type.getType());
	}

	@Override
	public String format(Object object) {
		return gson.toJson(object);
	}

	@Override
	public String formatUnicode(Object object) {
		return gsonUnicode.toJson(object);
	}

	@Override
	public String formatPretty(Object object) {
		return gsonPretty.toJson(object);
	}

	@Override
	public String formatPrettyUnicode(Object object) {
		return gsonPrettyUnicode.toJson(object);
	}

	@Override
	public boolean isValidJson(String json) {
		try {
			JsonElement element = JsonParser.parseString(json);
			return true;
		} catch (JsonSyntaxException e) {
			return false;
		} catch (Exception e) {
			return false;
		}
	}

	@Override
	public String merge(String... jsons) {
		JsonObject merged = new JsonObject();
		for (String json : jsons) {
			if (json == null || json.isEmpty()) {
				continue;
			}
			try {
				JsonObject obj = JsonParser.parseString(json).getAsJsonObject();
				for (String key : obj.keySet()) {
					merged.add(key, obj.get(key));
				}
			} catch (Exception e) {
				// 忽略无效的 JSON 字符串
			}
		}
		return gson.toJson(merged);
	}

	@Override
	public String getNodeValue(String json, String path) {
		try {
			JsonElement element = JsonParser.parseString(json);
			String[] paths = path.split("\\.");
			JsonElement current = element;

			for (String p : paths) {
				if (current.isJsonObject()) {
					current = current.getAsJsonObject().get(p);
				} else {
					return null;
				}

				if (current == null) {
					return null;
				}
			}

			return current.isJsonPrimitive()
				? current.getAsString()
				: current.toString();
		} catch (Exception e) {
			return null;
		}
	}

	@Override
	public String updateNodeValue(String json, String path, Object newValue) {
		try {
			JsonObject obj = JsonParser.parseString(json).getAsJsonObject();
			String[] paths = path.split("\\.");
			JsonObject current = obj;

			// 导航到倒数第二层
			for (int i = 0; i < paths.length - 1; i++) {
				String p = paths[i];
				if (!current.has(p) || !current.get(p).isJsonObject()) {
					current.add(p, new JsonObject());
				}
				current = current.getAsJsonObject(p);
			}

			// 设置最后一层的值
			String lastPath = paths[paths.length - 1];
			if (newValue == null) {
				current.remove(lastPath);
			} else {
				JsonElement element = gson.toJsonTree(newValue);
				current.add(lastPath, element);
			}

			return gson.toJson(obj);
		} catch (Exception e) {
			return json;
		}
	}

	@Override
	public <T, D> D convert(T source, Class<D> destinationClass) {
		String json = gson.toJson(source);
		return gson.fromJson(json, destinationClass);
	}

	@Override
	public <T, D> D convert(T source, JsonTypeReference<D> destinationType) {
		String json = gson.toJson(source);
		return gson.fromJson(json, destinationType.getType());
	}
}

File: src/main/kotlin/com/mingliqiye/utils/json/GsonJsonApi.kt
/*
 * Copyright 2025 mingliqiye
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * ProjectName mingli-utils
 * ModuleName mingli-utils.main
 * CurrentFile GsonJsonApi.kt
 * LastUpdate 2025-09-15 22:07:43
 * UpdateUser MingLiPro
 */

package com.mingliqiye.utils.json

import com.google.gson.*
import com.mingliqiye.utils.json.converters.JsonConverter
import com.mingliqiye.utils.json.converters.JsonStringConverter

class GsonJsonApi : JsonApi {

    private var gsonUnicode: Gson
    private var gsonPretty: Gson
    private var gsonPrettyUnicode: Gson
    private var gson: Gson

    constructor() {
        gson = GsonBuilder()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()

        gsonUnicode = GsonBuilder()
            .disableHtmlEscaping()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()

        gsonPretty = GsonBuilder()
            .setPrettyPrinting()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()

        gsonPrettyUnicode = GsonBuilder()
            .setPrettyPrinting()
            .disableHtmlEscaping()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()
    }

    constructor(gson: Gson) {
        this.gson = gson
            .newBuilder()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()
        this.gsonUnicode = gson
            .newBuilder()
            .disableHtmlEscaping()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()
        this.gsonPretty = gson
            .newBuilder()
            .setPrettyPrinting()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()
        this.gsonPrettyUnicode = gson
            .newBuilder()
            .setPrettyPrinting()
            .disableHtmlEscaping()
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .create()
    }

    override fun <T> parse(json: String, clazz: Class<T>): T {
        return gson.fromJson(json, clazz)
    }

    override fun <T> parse(json: String, type: JsonTypeReference<T>): T {
        return gson.fromJson(json, type.type)
    }

    override fun format(obj: Any): String {
        return gson.toJson(obj)
    }

    override fun formatUnicode(obj: Any): String {
        return gsonUnicode.toJson(obj)
    }

    override fun formatPretty(obj: Any): String {
        return gsonPretty.toJson(obj)
    }

    override fun formatPrettyUnicode(obj: Any): String {
        return gsonPrettyUnicode.toJson(obj)
    }

    override fun isValidJson(json: String): Boolean {
        return try {
            JsonParser.parseString(json)
            true
        } catch (e: JsonSyntaxException) {
            false
        } catch (e: Exception) {
            false
        }
    }

    override fun merge(vararg jsons: String): String {
        val merged = JsonObject()
        for (json in jsons) {
            if (json.isNullOrEmpty()) {
                continue
            }
            try {
                val obj = JsonParser.parseString(json).asJsonObject
                for (key in obj.keySet()) {
                    merged.add(key, obj.get(key))
                }
            } catch (e: Exception) {
                // 忽略无效的 JSON 字符串
            }
        }
        return gson.toJson(merged)
    }

    override fun getNodeValue(json: String, path: String): String? {
        return try {
            var element = JsonParser.parseString(json)
            val paths = path.split("\\.".toRegex()).toTypedArray()
            var current = element

            for (p in paths) {
                if (current.isJsonObject) {
                    current = current.asJsonObject.get(p)
                } else {
                    return null
                }

                if (current == null) {
                    return null
                }
            }

            if (current.isJsonPrimitive) current.asString else current.toString()
        } catch (e: Exception) {
            null
        }
    }

    override fun updateNodeValue(json: String, path: String, newValue: Any): String {
        return try {
            val obj = JsonParser.parseString(json).asJsonObject
            val paths = path.split("\\.".toRegex()).toTypedArray()
            var current = obj

            // 导航到倒数第二层
            for (i in 0 until paths.size - 1) {
                val p = paths[i]
                if (!current.has(p) || !current.get(p).isJsonObject) {
                    current.add(p, JsonObject())
                }
                current = current.getAsJsonObject(p)
            }

            // 设置最后一层的值
            val lastPath = paths[paths.size - 1]
            val element = gson.toJsonTree(newValue)
            current.add(lastPath, element)

            gson.toJson(obj)
        } catch (e: Exception) {
            json
        }
    }

    override fun <T, D> convert(source: T, destinationClass: Class<D>): D {
        val json = gson.toJson(source)
        return gson.fromJson(json, destinationClass)
    }

    override fun <T, D> convert(source: T, destinationType: JsonTypeReference<D>): D {
        val json = gson.toJson(source)
        return gson.fromJson(json, destinationType.type)
    }

    override fun addJsonConverter(c: JsonConverter<*, *>) {
        c.getStringConverter()?.let {
            gson = gson
                .newBuilder()
                .registerTypeAdapter(
                    it.tClass,
                    it.gsonJsonStringConverterAdapter
                )
                .create()
            gsonUnicode = gsonUnicode
                .newBuilder()
                .registerTypeAdapter(
                    it.tClass,
                    it.gsonJsonStringConverterAdapter
                )
                .create()
            gsonPretty = gsonPretty
                .newBuilder()
                .registerTypeAdapter(
                    it.tClass,
                    it.gsonJsonStringConverterAdapter
                )
                .create()
            gsonPrettyUnicode = gsonPrettyUnicode
                .newBuilder()
                .registerTypeAdapter(
                    it.tClass,
                    it.gsonJsonStringConverterAdapter
                )
                .create()
        }

    }

    override fun addJsonStringConverter(c: JsonStringConverter<*>) {
        addJsonConverter(c)
    }
}
